components/wmtk_components/marching/internal/Marching.cpp

```diff
@@ -0,0 +1,112 @@
+
+#include <wmtk/SimplicialComplex.hpp>
+#include <wmtk/operations/tri_mesh/EdgeCollapseToMidpoint.hpp>
+#include <wmtk/operations/tri_mesh/EdgeSplitAtMidpoint.hpp>
+#include <wmtk/operations/tri_mesh/EdgeSplitWithTag.hpp>
+#include <wmtk/operations/tri_mesh/EdgeSwap.hpp>
+#include <wmtk/operations/tri_mesh/FaceSplitWithTag.hpp>
+#include <wmtk/operations/tri_mesh/VertexTangentialLaplacianSmooth.hpp>
+
+namespace wmtk::components::internal {
+
+Marching::Marching(
+    TriMesh& mesh,
+    MeshAttributeHandle<double>& position_handle,
+    MeshAttributeHandle<long>& vertex_tag,
+    MeshAttributeHandle<long>& edge_tag,
+    const long input_tag_value,
+    const long embedding_tag_value,
+    const long split_tag_value,
+    const int dimension,
+    const bool lock_boundary)
+    : m_mesh(mesh)
+    , m_position_handle(position_handle)
+    , m_vertex_tag(vertex_tag)
+    , m_edge_tag(edge_tag)
+    , m_input_tag_value(input_tag_value)
+    , m_embedding_tag_value(embedding_tag_value)
+    , m_split_tag_value(split_tag_value)
+    , m_dimension(dimension)
+    , m_scheduler(mesh)
+    , m_lock_boundary(lock_boundary)
+{}
+
+void Marching::process()
+{
+    switch (m_dimension) {
+    case 2: process_in_2d(); break;
+    case 3: process_in_3d(); break;
+    default: throw std::runtime_error("settings went wrong!"); break;
+    }
+}
+
+void Marching::process_in_2d()
+{
+    using namespace operations;
+
+    wmtk::MeshAttributeHandle<long> todo_edgesplit_same_handle = m_mesh.register_attribute<long>(
+        std::string("todo_edgesplit_same_tag"),
+        wmtk::PrimitiveType::Edge,
+        1);
+
+    wmtk::Accessor<long> acc_vertex_tag = m_mesh.create_accessor(m_vertex_tag);
+    wmtk::Accessor<long> acc_edge_tag = m_mesh.create_accessor(m_edge_tag);
+    wmtk::Accessor<double> acc_pos = m_mesh.create_accessor(m_position_handle);
+    wmtk::Accessor<long> acc_todo_edgesplit_same_tag =
+        m_mesh.create_accessor(todo_edgesplit_same_handle);
+
+    // edge split
+    {
+        // compute the todo list for the split edge with the same ends
+        const std::vector<Tuple>& edges = m_mesh.get_all(wmtk::PrimitiveType::Edge);
+        for (const Tuple& edge : edges) {
+            long vt0, vt1, et0;
+            vt0 = acc_vertex_tag.scalar_attribute(edge);
+            vt1 = acc_vertex_tag.scalar_attribute(m_mesh.switch_vertex(edge));
+            if ((vt0 == m_input_tag_value && vt1 == m_embedding_tag_value) ||
+                (vt1 == m_input_tag_value && vt0 == m_embedding_tag_value)) {
+                acc_todo_edgesplit_same_tag.scalar_attribute(edge) = 1;
+            }
+        }
+        // using scheduler to do edge splitting
+        OperationSettings<tri_mesh::EdgeSplitWithTag> settings_split;
+        settings_split.edge_tag = m_edge_tag;
+        settings_split.vertex_tag = m_vertex_tag;
+        settings_split.embedding_tag_value = m_embedding_tag_value;
+        settings_split.need_embedding_tag_value = true;
```

Comment:
What is that setting? It sounds oddly specific.

Reply:
I will make a comment here. when you do a split operation, the new four edges could be tagged with a specific value. By default, they will be tagged as embedding values. but I think sometimes we don't want to change the original edge's tag, so this is where the need_embedding_tag_value comes from.